includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -258,29 +277,125 @@ public function sanitize() {
+			if ( $node instanceof AMP_Contextual_Node ) {
+				array_shift( $this->stack );
+				continue;
+			}
+
+			// No need to proceed further if the node has no children.
+			if ( ! $node->parentNode ) {
+				array_shift( $this->stack );
+				continue;
+			}
+
 			/*
 			 * Push child nodes onto the stack, if any exist.
 			 * if node was removed, then it's parentNode value is null.
 			 */
 			if ( $node->parentNode ) {
-				$child = $node->firstChild;
-				while ( $child ) {
-					$this->stack[] = $child;
-					$child         = $child->nextSibling;
+
+				$child_nodes = $this->has_child_nodes( $node );
+
+				// If we need contextual nodes to track state for the current element,
+				// we add them as wrapping nodes around all children.
+				$contextual_nodes = $this->needs_contextual_nodes( $node );
+
+				if ( ! empty( $contextual_nodes ) ) {
+					array_unshift( $child_nodes, array_shift( $contextual_nodes ) );
+					array_push( $child_nodes, array_shift( $contextual_nodes ) );
 				}
+
+				// Replace the parent node by its child nodes.
+				array_splice( $this->stack, $node_index, 1, $child_nodes );
 			}
 		}
 	}
 
+	/**
+	 * Check whether a given node requires keeping track of context.
+	 *
+	 * If this is the case, it will return a pair of start and end nodes to
+	 * keep track of said context. This is done because we linearize the stack
+	 * and lack hierarchical information when processing the nodes. So we turn
+	 * one parent node into a pair of starting and ending contextual nodes.
+	 *
+	 * @param DOMNode $node Node to examine.
+	 * @return DOMNode[]|false Pair of contextual nodes, or false if none to be
+	 *                         used.
+	 */
+	protected function needs_contextual_nodes( DOMNode $node ) {
+		switch ( $node->nodeName ) {
+			case 'template':
+				return $this->get_template_contextual_nodes( $node );
+		}
+
+		return false;
+	}
+
+	/**
+	 * Retrieve an array of direct child nodes from a given node.
+	 *
+	 * @param DOMNode $node Node to retrieve the direct children from.
+	 * @return DOMNode[] Array of child nodes.
+	 */
+	protected function has_child_nodes( DOMNode $node ) {
+		$children = [];
+		$child    = $node->firstChild;
+
+		while ( $child ) {
+			$children[] = $child;
+			$child      = $child->nextSibling;
+		}
+
+		return $children;
+	}
+
+	/**
+	 * Build and return the contextual nodes needed for template tags.
+	 *
+	 * @param DOMNode $node Template tag to examine.
+	 * @return DOMNode[]|false Pair of contextual nodes, or false if none to be
+	 *                         used.
+	 */
+	protected function get_template_contextual_nodes( DOMNode $node ) {
+		// A template tag is only valid if it has a type attribute.
+		if ( ! $node->hasAttributes() ) {
+			return false;
+		}
+		$type = $node->attributes->getNamedItem( 'type' );
+		if ( null === $type ) {
+			return false;
+		}
+
+		// The type attribute of a template tag defines what templating engine to use.
+		switch ( $type->nodeValue ) {
```

Comment:
This `switch` only has one case so far, but this properly communicates the intent that there are multiple possible values we need to consider here. Turning it into an `if` for the short-term would obfuscate this intent and probably lead to problems later on when extending.